File: src/main/java/org/kobe/xbot/JClient/SubscribeHandler.java
package org.kobe.xbot.JClient;

import com.google.protobuf.ByteString;
import org.kobe.xbot.Utilities.CircularBuffer;
import org.kobe.xbot.Utilities.ClientStatistics;
import org.kobe.xbot.Utilities.Entities.XTableClientStatistics;
import org.kobe.xbot.Utilities.Entities.XTableProto;
import org.kobe.xbot.Utilities.Exceptions.XTablesException;
import org.kobe.xbot.Utilities.Utilities;
import org.zeromq.ZMQ;

import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.function.Consumer;

/**
 * SubscribeHandler - A handler for processing incoming subscription messages using JeroMQ.
 * <p>
 * This class handles the reception of messages on a ZeroMQ socket, processes them, and stores the updates in a circular buffer.
 * It also manages a separate consumer thread that handles the subscription consumers and invokes their functions for each new update.
 * The handler supports both UPDATE and PUBLISH message categories.
 * <p>
 * Author: Kobe Lei
 * Version: 1.0
 * Package: org.kobe.xbot.JClient
 * <p>
 * This is part of the XTABLES project and facilitates subscribing and handling incoming messages from the server.
 */
public class SubscribeHandler extends BaseHandler {
    private final XTablesClient instance;
    private final CircularBuffer<XTableProto.XTableMessage.XTableUpdate> buffer;
    private final Thread consumerHandlingThread;
    private final static int BUFFER_SIZE = 500;
    private final SubscriberManager subscriberManager;
    /**
     * Constructor that initializes the handler with the provided socket and server instance.
     *
     * @param socket   The ZeroMQ socket to receive messages on
     * @param instance The XTablesClient instance
     */
    public SubscribeHandler(ZMQ.Socket socket, XTablesClient instance) {
        super("XTABLES-SUBSCRIBE-HANDLER-DAEMON", true, socket);
        this.instance = instance;
        this.buffer = new CircularBuffer<>(BUFFER_SIZE, (latest, current) -> current.getKey().equals(latest.getKey()));
        this.consumerHandlingThread = new ConsumerHandlingThread();
        this.consumerHandlingThread.start();
        this.subscriberManager = new SubscriberManager(socket);
        this.subscriberManager.start();
    }

    /**
     * Requests a new subscription to a topic.
     *
     * @param topic The topic to subscribe to.
     */
    public boolean requestSubscribe(byte[] topic) {
       return subscriberManager.requestSubscription(topic);
    }

    /**
     * Requests unsubscription from a topic.
     *
     * @param topic The topic to unsubscribe from.
     */
    public boolean requestUnsubscription(byte[] topic) {
       return subscriberManager.requestUnsubscription(topic);
    }
    /**
     * The main method for handling incoming messages.
     * <p>
     * It continuously receives messages from the JeroMQ socket,
     * processes them by parsing the message into a protocol buffer,
     * and stores the updates in the buffer.
     * It also increments the count of subscription messages received.
     */
    @Override
    public void run() {
        try {
            while (!Thread.currentThread().isInterrupted()) {
                byte[] bytes = socket.recv();
                try {
                    XTableProto.XTableMessage.XTableUpdate message = XTableProto.XTableMessage.XTableUpdate.parseFrom(bytes);
                    if (message.getCategory().equals(XTableProto.XTableMessage.XTableUpdate.Category.INFORMATION) || message.getCategory().equals(XTableProto.XTableMessage.XTableUpdate.Category.REGISTRY)) {
                        byte[] info = new ClientStatistics()
                                .setBufferSize(buffer.size)
                                .setUUID(XTablesClient.UUID)
                                .setVersion(instance.getVersion())
                                .setMaxBufferSize(BUFFER_SIZE)
                                        .toProtobuf()
                                                .toByteArray();

                        instance.getRegsitrySocket().send(XTableProto.XTableMessage.newBuilder()
                                .setId(message.getValue())
                                .setValue(ByteString.copyFrom(info))
                                .setCommand(message.getCategory().equals(XTableProto.XTableMessage.XTableUpdate.Category.INFORMATION) ? XTableProto.XTableMessage.Command.INFORMATION : XTableProto.XTableMessage.Command.REGISTRY)
                                .build().toByteArray(), ZMQ.DONTWAIT);
                    } else this.buffer.write(message);
                } catch (Exception e) {
                    handleException(e);
                }
            }
        } catch (Exception e) {
            handleException(e);
        }
    }

    /**
     * Logs exceptions and handles cleanup when the thread is interrupted.
     * <p>
     * This method interrupts the consumer handling thread and calls the parent class's interrupt method to clean up.
     */
    @Override
    public void interrupt() {
        this.consumerHandlingThread.interrupt();
        this.subscriberManager.interrupt();
        super.interrupt();
    }

    /**
     * ConsumerHandlingThread - A thread for handling subscription consumers for each received update.
     * <p>
     * This inner thread continuously reads the latest update from the buffer and invokes the consumers
     * that are subscribed to the specific update category or key.
     * It handles both general and registry updates.
     */
    private class ConsumerHandlingThread extends Thread {
        public ConsumerHandlingThread() {
            setName("XTABLES-CONSUMER-HANDLER-DAEMON");
            setDaemon(true);
        }

        /**
         * The run method processes each update by invoking the appropriate consumers.
         * <p>
         * It checks if the update is of type UPDATE, PUBLISH, or REGISTRY. For UPDATE and PUBLISH, it invokes all consumers
         * associated with the update key. For REGISTRY, it sends the registry command to the push socket.
         */
        @Override
        public void run() {
            try {
                while (!Thread.currentThread().isInterrupted()) {
                    XTableProto.XTableMessage.XTableUpdate update = buffer.readLatestAndClearOnFunction();
                    if (update.getCategory().equals(XTableProto.XTableMessage.XTableUpdate.Category.UPDATE) || update.getCategory().equals(XTableProto.XTableMessage.XTableUpdate.Category.PUBLISH)) {
                        if (instance.subscriptionConsumers.containsKey(update.getKey())) {
                            List<Consumer<XTableProto.XTableMessage.XTableUpdate>> consumers = instance.subscriptionConsumers.get(update.getKey());

                            for (Consumer<XTableProto.XTableMessage.XTableUpdate> consumer : consumers) {
                                consumer.accept(update);
                            }
                        }
                        if (instance.subscriptionConsumers.containsKey("")) {
                            List<Consumer<XTableProto.XTableMessage.XTableUpdate>> consumers = instance.subscriptionConsumers.get("");
                            for (Consumer<XTableProto.XTableMessage.XTableUpdate> consumer : consumers) {
                                consumer.accept(update);
                            }
                        }
                    } else if (update.getCategory().equals(XTableProto.XTableMessage.XTableUpdate.Category.LOG)) {
                        for (Consumer<XTableProto.XTableMessage.XTableLog> consumer : instance.logConsumers) {
                            consumer.accept(XTableProto.XTableMessage.XTableLog.parseFrom(update.getValue()));
                        }
                    }
                }
            } catch (Exception e) {
                handleException(new XTablesException(e));
            }
        }
    }

    /**
     * SubscriberManager - Manages the ZeroMQ SUB socket subscriptions safely in a single thread.
     */
    private class SubscriberManager extends Thread {
        private final ZMQ.Socket subscriber;
        private final BlockingQueue<byte[]> subscriptionQueue;
        private final BlockingQueue<byte[]> unsubscriptionQueue;

        public SubscriberManager(ZMQ.Socket subscriber) {
            this.subscriber = subscriber;
            this.subscriptionQueue = new LinkedBlockingQueue<>();
            this.unsubscriptionQueue = new LinkedBlockingQueue<>();
        }

        public boolean requestSubscription(byte[] topic) {
           return subscriptionQueue.offer(topic);
        }

        public boolean requestUnsubscription(byte[] topic) {
           return unsubscriptionQueue.offer(topic);
        }

        @Override
        public void run() {
            while (!Thread.currentThread().isInterrupted()) {
                try {
                    // Process subscriptions
                    byte[] topic;
                    while ((topic = subscriptionQueue.poll()) != null) {
                        if(!subscriber.subscribe(topic)) {
                            subscriptionQueue.put(topic);
                        }
                    }

                    // Process unsubscriptions
                    while ((topic = unsubscriptionQueue.poll()) != null) {
                        if(!subscriber.unsubscribe(topic)) {
                            unsubscriptionQueue.put(topic);
                        };
                    }

                    Thread.sleep(100); // Prevent CPU overuse

                } catch (InterruptedException e) {
                    break;
                } catch (Exception e) {
                    handleException(e);
                }
            }
            subscriber.close();
        }
    }
}

File: src/main/python/JXTABLES/SubscribeHandler.py
import threading
from typing import Callable, List, Dict, Any
import zmq
import traceback
from google.protobuf.message import DecodeError

try:
    # Package-level imports
    from .BaseHandler import BaseHandler
    from .ClientStatistics import ClientStatistics
    from .CircularBuffer import CircularBuffer
    from . import Utilities
    from . import XTableProto_pb2 as XTableProto
except ImportError:
    # Standalone script imports
    from BaseHandler import BaseHandler
    from ClientStatistics import ClientStatistics
    from CircularBuffer import CircularBuffer
    import Utilities
    import XTableProto_pb2 as XTableProto


class SubscribeHandler(BaseHandler):
    """
    A handler for processing incoming subscription messages using ZeroMQ.
    """

    def __init__(self, socket: zmq.Socket, instance: Any):
        super().__init__("XTABLES-SUBSCRIBE-HANDLER-DAEMON", True, socket)
        self._stop = False
        self.BUFFER_SIZE = instance.BUFFER_SIZE
        self.instance = instance
        self.buffer = CircularBuffer(self.BUFFER_SIZE, lambda latest, current: latest.key == current.key)
        self.consumer_handling_thread = self.ConsumerHandlingThread(self)
        self.consumer_handling_thread.start()

    def run(self):
        """
        Main loop for handling incoming messages.
        """
        try:
            while not self._stop:
                try:
                    bytes_message = self.socket.recv()
                    message = XTableProto.XTableMessage.XTableUpdate.FromString(bytes_message)

                    if message.category in {XTableProto.XTableMessage.XTableUpdate.Category.INFORMATION,
                                            XTableProto.XTableMessage.XTableUpdate.Category.REGISTRY}:
                        if not self.instance.ghost:
                            stats = ClientStatistics()
                            stats.set_buffer_size(self.buffer.size)
                            stats.set_uuid(self.instance.uuid)
                            stats.set_version(self.instance.get_version())
                            stats.set_max_buffer_size(self.BUFFER_SIZE)
                            info_bytes = stats.to_protobuf()
                            response = XTableProto.XTableMessage(
                                id=message.value,
                                value=info_bytes.SerializeToString(),
                                command=XTableProto.XTableMessage.Command.INFORMATION
                                if message.category == XTableProto.XTableMessage.XTableUpdate.Category.INFORMATION
                                else XTableProto.XTableMessage.Command.REGISTRY,
                            )
                            self.instance.registry_socket.send(response.SerializeToString(), zmq.DONTWAIT)
                    else:
                        self.buffer.write(message)

                except DecodeError:
                    if self.instance.debug:
                        traceback.print_exc()
                except Exception as e:
                    if self.instance.debug:
                        traceback.print_exc()
                    self.handle_exception(e)
        except Exception as e:
            if self.instance.debug:
                traceback.print_exc()
            self.handle_exception(e)

    def interrupt(self):
        """
        Handles cleanup and interrupts the consumer handling thread.
        """
        self._stop = True
        super().interrupt()

    class ConsumerHandlingThread(threading.Thread):
        """
        A thread for handling subscription consumers for each received update.
        """

        def __init__(self, parent_handler: 'SubscribeHandler'):
            super().__init__(name="XTABLES-CONSUMER-HANDLER-DAEMON", daemon=True)
            self.parent_handler = parent_handler

        def run(self):
            """
            Continuously processes updates from the buffer and invokes subscribed consumers.
            """
            try:
                while not self.parent_handler._stop:
                    try:
                        update = self.parent_handler.buffer.read_latest_and_clear_on_function()

                        if update.category in {XTableProto.XTableMessage.XTableUpdate.Category.UPDATE,
                                               XTableProto.XTableMessage.XTableUpdate.Category.PUBLISH}:
                            self._process_consumers(update)
                        elif update.category == XTableProto.XTableMessage.XTableUpdate.Category.LOG:
                            self._process_log_consumers(update)

                    except Exception as e:
                        if self.parent_handler.instance.debug:
                            traceback.print_exc()
                        self.parent_handler.handle_exception(e)
            except Exception as e:
                if self.parent_handler.instance.debug:
                    traceback.print_exc()
                self.parent_handler.handle_exception(e)

        def _process_consumers(self, update):
            """
            Process consumers for UPDATE and PUBLISH categories.
            """
            if update.key in self.parent_handler.instance.subscription_consumers:
                consumers = self.parent_handler.instance.subscription_consumers[update.key]
                for consumer in consumers:
                    consumer(update)

            if "" in self.parent_handler.instance.subscription_consumers:
                general_consumers = self.parent_handler.instance.subscription_consumers[""]
                for consumer in general_consumers:
                    consumer(update)

        def _process_log_consumers(self, update):
            """
            Process log consumers for LOG category.
            """
            for consumer in self.parent_handler.instance.log_consumers:
                log_message = XTableProto.XTableMessage.XTableLog.FromString(update.value)
                consumer(log_message)
